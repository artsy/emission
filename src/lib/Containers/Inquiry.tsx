import { Inquiry_artwork } from "__generated__/Inquiry_artwork.graphql"
import colors from "lib/data/colors"
import fonts from "lib/data/fonts"
import React from "react"
import { Dimensions, NativeModules } from "react-native"
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components/native"
import { BottomAlignedButton } from "../Components/Consignments/Components/BottomAlignedButton"
import ArtworkPreview from "../Components/Inbox/Conversations/Preview/ArtworkPreview"
import { MetadataText, SmallHeadline } from "../Components/Inbox/Typography"
import ARSwitchBoard from "../NativeModules/SwitchBoard"
import { gravityURL } from "../relay/config"
import { NetworkError } from "../utils/errors"
import { Schema, Track, track as _track } from "../utils/track"

const isPad = Dimensions.get("window").width > 700

const Container = styled.View`
  flex: 1;
  flex-direction: column;
  background-color: white;
`
const Header = styled.View`
  align-self: stretch;
  margin-top: 10;
  flex-direction: column;
  margin-bottom: 30;
`
// This is really rubbish, but I basically have to create an equally sized element
// on the top right, to get the title in the middle
const PlaceholderView = styled(SmallHeadline)`
  padding-right: 20;
  color: white;
`
const TitleView = styled.View`
  align-self: center;
  align-items: center;
  margin-top: 6;
`
const PartnerName = styled(SmallHeadline)`
  font-size: 12;
`
const HeaderTextContainer = styled.View`
  flex-direction: row;
  justify-content: space-between;
`
const CancelButton = styled.TouchableOpacity`
  padding-left: 20;
`

const Content = styled.View`
  margin-left: 20;
  margin-right: 20;
  align-self: ${isPad ? "center" : "stretch"};
  ${isPad ? "width: 472;" : ""};
`

const InquiryTextInput = styled.TextInput`
  font-size: 16;
  margin-top: 20;
  font-family: ${fonts["garamond-regular"]};
`
const ResponseRate = styled(SmallHeadline)`
  color: ${colors["yellow-bold"]};
  margin-top: 5;
`
// TODO: Uncomment when use is uncommented in code below
// const ResponseIndicator = styled.View`
//   width: 8;
//   height: 8;
//   border-radius: 4;
//   margin-top: 5;
//   margin-right: 5;
//   background-color: ${colors["yellow-bold"]};
// `

const ResponseRateLine = styled.View`
  flex: 1;
  flex-direction: row;
  align-items: center;
  min-height: 12;
  margin-top: 5;
`

interface Props {
  artwork: Inquiry_artwork
}

interface State {
  text: string
  sending: boolean
}

const track: Track<Props, State, Schema.Entity> = _track

@track()
export class Inquiry extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      text: this.props.artwork.contact_message,
      sending: false,
    }
  }

  @track(props => ({
    action_type: Schema.ActionTypes.Tap,
    action_name: Schema.ActionNames.InquiryCancel,
    owner_type: Schema.OwnerEntityTypes.Artwork,
    owner_id: props.artwork.internalID,
    owner_slug: props.artwork.slug,
  }))
  cancelModal() {
    this.dismissModal()
  }

  @track(props => ({
    action_type: Schema.ActionTypes.Success,
    action_name: Schema.ActionNames.InquirySend,
    owner_type: Schema.OwnerEntityTypes.Artwork,
    owner_id: props.artwork.internalID,
    owner_slug: props.artwork.slug,
  }))
  inquirySent() {
    this.dismissModal()
  }

  dismissModal() {
    ARSwitchBoard.dismissModalViewController(this)
  }

  @track(props => ({
    action_type: Schema.ActionTypes.Tap,
    action_name: Schema.ActionNames.InquirySend,
    owner_type: Schema.OwnerEntityTypes.Artwork,
    owner_id: props.artwork.internalID,
    owner_slug: props.artwork.slug,
  }))
  sendInquiry() {
    // Using setState to trigger re-render for the button
    this.setState(() => ({ sending: true }))
    const { Emission } = NativeModules
    fetch(gravityURL + "/api/v1/me/artwork_inquiry_request", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-ACCESS-TOKEN": Emission.authenticationToken,
        "User-Agent": Emission.userAgent,
      },
      body: JSON.stringify({
        artwork: this.props.artwork.slug,
        message: this.state.text,
      }),
    })
      .then(response => {
        if (response.status >= 200 && response.status < 300) {
          this.inquirySent()
        } else {
          const error = new NetworkError(response.statusText)
          error.response = response
          this.sendFailed(error)
        }
      })
      .catch(error => {
        this.sendFailed(error)
      })
  }

  @track(props => ({
    action_type: Schema.ActionTypes.Fail,
    action_name: Schema.ActionNames.InquirySend,
    owner_type: Schema.OwnerEntityTypes.Artwork,
    owner_id: props.artwork.internalID,
    owner_slug: props.artwork.slug,
  }))
  sendFailed(error) {
    this.setState(() => ({ sending: false }))
    throw error
  }

  render() {
    const message = this.state.text
    const partnerResponseRate = " " // currently hardcoded for alignment
    const artwork = this.props.artwork
    const partnerName = this.props.artwork.partner.name
    const buttonText = this.state.sending ? "Sending..." : "Send"

    const doneButtonStyles = {
      backgroundColor: colors["purple-regular"],
      marginBottom: 0,
      paddingTop: 15,
      height: 50,
    }

    return (
      <Container>
        <BottomAlignedButton
          onPress={this.sendInquiry.bind(this)}
          bodyStyle={doneButtonStyles}
          buttonText={buttonText}
          disabled={this.state.sending}
        >
          <Header>
            <HeaderTextContainer>
              <CancelButton onPress={this.cancelModal.bind(this)}>
                <MetadataText>Cancel</MetadataText>
              </CancelButton>
              <TitleView>
                <PartnerName>{partnerName}</PartnerName>
                <ResponseRateLine>
                  {/* <ResponseIndicator /> */}
                  <ResponseRate>{partnerResponseRate}</ResponseRate>
                </ResponseRateLine>
              </TitleView>
              <PlaceholderView>Cancel</PlaceholderView>
            </HeaderTextContainer>
          </Header>
          <Content>
            <ArtworkPreview artwork={artwork as any} />
            <InquiryTextInput
              value={message}
              keyboardAppearance="dark"
              multiline={true}
              autoFocus={typeof jest === "undefined" /* TODO: https://github.com/facebook/jest/issues/3707 */}
              onEndEditing={() => {
                this.setState({ text: null })
              }}
              onChangeText={text => this.setState({ text })}
            />
          </Content>
        </BottomAlignedButton>
      </Container>
    )
  }
}

export default createFragmentContainer(Inquiry, {
  artwork: graphql`
    fragment Inquiry_artwork on Artwork {
      slug
      internalID
      contact_message
      partner {
        name
      }
      ...ArtworkPreview_artwork
    }
  `,
})
