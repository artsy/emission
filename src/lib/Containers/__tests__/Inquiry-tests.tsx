import * as React from "react"
import "react-native"
import * as renderer from "react-test-renderer"
import Inquiry from "../Inquiry"

it("renders correctly", () => {
  const tree = renderer.create(<Inquiry artwork={inquiryProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})

const inquiryProps = {
  id: "bradley-theodore-karl-and-anna-face-off-diptych",
  contact_message:
    "Hi, I'm interested in purchasing this work. Could you please provide me with more information about the piece?",
  href: "/artwork/bradley-theodore-karl-and-anna-face-off-diptych",
  title: "Karl and Anna Face Off (Diptych)",
  date: "2016",
  artist_names: "Bradley Theodore",
  image: {
    url: "https://d32dm0rphc51dk.cloudfront.net/bJ9I_vJX9ksaKFJAkOAIKg/normalized.jpg",
  },
  partner: {
    name: "Other Criteria",
  },
}
