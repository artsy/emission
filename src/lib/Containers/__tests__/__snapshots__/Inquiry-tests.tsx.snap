// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "flexGrow": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "flexDirection": "column",
              "marginBottom": 30,
              "marginTop": 10,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
                "paddingLeft": 20,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#cccccc",
                    "fontSize": 11,
                    "textAlign": "left",
                  },
                  Object {},
                  Object {
                    "fontFamily": "AvantGardeGothicITC",
                  },
                ]
              }
            >
              CANCEL
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "marginTop": 6,
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "black",
                    "fontSize": 11,
                    "textAlign": "left",
                  },
                  undefined,
                  Array [
                    Object {
                      "fontSize": 12,
                    },
                    undefined,
                  ],
                  Object {
                    "fontFamily": "AvantGardeGothicITC",
                  },
                ]
              }
            >
              OTHER CRITERIA
            </Text>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexBasis": 0,
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "marginTop": 5,
                    "minHeight": 12,
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "black",
                      "fontSize": 11,
                      "textAlign": "left",
                    },
                    undefined,
                    Array [
                      Object {
                        "color": "#f1af1b",
                        "marginTop": 5,
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": "AvantGardeGothicITC",
                    },
                  ]
                }
              >
                 
              </Text>
            </View>
          </View>
          <Text
            style={
              Array [
                Object {
                  "color": "black",
                  "fontSize": 11,
                  "textAlign": "left",
                },
                undefined,
                Array [
                  Object {
                    "color": "white",
                    "paddingRight": 20,
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "AvantGardeGothicITC",
                },
              ]
            }
          >
            CANCEL
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "marginLeft": 20,
              "marginRight": 20,
              "width": 472,
            },
            undefined,
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderWidth": 1,
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <AROpaqueImageView
              imageURL="https://d32dm0rphc51dk.cloudfront.net/bJ9I_vJX9ksaKFJAkOAIKg/normalized.jpg"
              style={
                Array [
                  Object {
                    "height": 55,
                    "marginBottom": 12,
                    "marginLeft": 12,
                    "marginRight": 12,
                    "marginTop": 12,
                    "width": 80,
                  },
                  undefined,
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "flexBasis": 0,
                    "flexDirection": "column",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                  Array [
                    Object {
                      "alignSelf": "center",
                    },
                    undefined,
                  ],
                ]
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "black",
                      "fontSize": 16,
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 14,
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": "AGaramondPro-Regular",
                    },
                  ]
                }
              >
                Bradley Theodore
              </Text>
              <View
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                      "marginRight": 12,
                      "marginTop": 3,
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  ellipsizeMode="middle"
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "black",
                        "fontSize": 14,
                      },
                      Object {},
                      Object {
                        "fontFamily": "AGaramondPro-Italic",
                      },
                    ]
                  }
                >
                  Karl and Anna Face Off (Diptych)
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "AGaramondPro-Regular",
                        },
                        undefined,
                      ]
                    }
                  >
                    , 2016
                  </Text>
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          autoFocus={false}
          keyboardAppearance="dark"
          multiline={true}
          onChangeText={[Function]}
          onEndEditing={[Function]}
          style={
            Array [
              Object {
                "fontFamily": "AGaramondPro-Regular",
                "fontSize": 16,
                "marginTop": 20,
              },
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value="Hi, I'm interested in purchasing this work. Could you please provide me with more information about the piece?"
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#666666",
            "height": 1,
          },
          undefined,
        ]
      }
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#6e1fff",
          "height": 50,
          "marginBottom": 0,
          "opacity": 1,
          "paddingTop": 15,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontFamily": "AvantGardeGothicITC",
              "fontSize": 14,
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        SEND
      </Text>
    </View>
  </View>
</View>
`;
